src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -502,6 +518,9 @@ private void ReadCentralDirectory()
                     numberOfEntries++;
                 }
 
+                // Sort _entries by each archive entry's position
+                _entries.Sort(ZipArchiveEntry.LocalHeaderOffsetComparer.Instance);
```

Comment:
Thanks for your review.

This comes partially from the [spec](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT). Section 4.3.2 says:

>4.3.2 Each file placed into a ZIP file MUST be preceded by  a "local file header" record for that file.  Each "local file header" MUST be accompanied by a corresponding "central directory header" record within the central directory section of the ZIP file.

Section 4.4.1.3 says:
>4.4.1.3  The entries in the central directory MAY NOT necessarily be in the same order that files appear in the .ZIP file.

Without the sort, `_entries` will be in the order of the central directory headers within the ZIP file. This works well, as long as the order of the central directory headers matches the order of the local file headers. The sort is designed to handle cases where the ZIP archive could contain entries in the sequence A, B, C, and central directory headers in the sequence A, C, B.

Reply:
Thanks for sharing the part of the spec indicating that order cannot be guaranteed.

Still, sorting them would be outside of what the spec indicates. We need to make sure we won't be causing a major perf decrease in cases where we modify zip archives that have been heavily updated.

Some things I would ask to ease our minds:

- Have you seen other zip APIs that sort their entries before iterating them?
- Have you tested this opening a file that has most or all their entries in the wrong order?
- Did you get a chance to run the unit tests locally and see if they were taking longer after this change?

If you can create a big zip file with most or all the entries in the wrong order and then share with me, I can help benchmark it and see if there's a big hit.

Reply:
I've benchmarked the cost of the sort and taken another look at the point in general.

I've put the full benchmark results below, but these essentially amount to 1.7us for 100 files, 133.8us for 5,000 files, 272us for 10,000 files and 1.68ms for 50,000 files. I didn't run the tests locally after the change, but these results would be too brief to actually change the durations.

More broadly though: we sort the ZipArchiveEntrys by their local file header offset because `WriteFile`'s algorithm for figuring out the right place to start writing to the stream relies on the items being in that order. It's possible to do that without the sort, but I think it'd take multiple iterations over the items - it needs to know which entries follow the modified entry in the ZipArchive. I've not seen any libraries which sort like this directly, although SharpZipLib comes quite close in [`ZipFile.BeginUpdate`](https://github.com/icsharpcode/SharpZipLib/blob/master/src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs#L1602); it sorts the list of archive entry updates, not the entire list of entries in the archive.

For performance purposes, I could potentially only perform the sort if the ZipArchive is in Update mode. There are some cases where users would see entries appear in a different order depending on whether they opened it in Read or Update mode, but that might not be a problem - they're usually going to open it by name.

I could technically also shift the sort into the WriteFile method, only performing it if the ZipArchive is being updated. I'm slightly reluctant to do that because I can see an API proposal for ZipArchive.Flush; sorting the list would mean that the order of entries in the archive would change after the user flushes.

I don't think we've got a unit test where the order of central directory entries is different to the order of local file headers, I'll add that over the next few days.

<details>
<summary>LocalHeaderOffsetComparer performance</summary>

```
BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4460/23H2/2023Update/SunValley3)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.100-rc.1.24452.12
  [Host]   : .NET 9.0.0 (9.0.24.43107), X64 RyuJIT AVX2
  Baseline : .NET 9.0.0 (9.0.24.43107), X64 RyuJIT AVX2

Job=Baseline  EnvironmentVariables=Empty  Toolchain=.NET 9.0  
InvocationCount=65536  UnrollFactor=16
```
| Method    | NumberOfFiles | RandomOrder | Mean         | Error     | StdDev    | Median       | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|---------- |-------------- |------------ |-------------:|----------:|----------:|-------------:|------:|--------:|-------:|-------:|----------:|------------:|
| Benchmark | 100           | False       |     1.703 μs | 0.0416 μs | 0.1193 μs |     1.717 μs |  1.01 |    0.10 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 100           | True        |     1.696 μs | 0.0399 μs | 0.1125 μs |     1.697 μs |  1.00 |    0.09 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 5000          | False       |   133.825 μs | 2.0884 μs | 1.7439 μs |   133.406 μs |  1.00 |    0.02 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 5000          | True        |   129.090 μs | 2.5411 μs | 4.5821 μs |   126.966 μs |  1.00 |    0.05 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 10000         | False       |   271.985 μs | 4.6164 μs | 4.3182 μs |   270.526 μs |  1.00 |    0.02 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 10000         | True        |   275.864 μs | 5.3996 μs | 7.7440 μs |   271.551 μs |  1.00 |    0.04 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 50000         | False       | 1,681.595 μs | 8.1089 μs | 7.1883 μs | 1,682.396 μs |  1.00 |    0.01 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 50000         | True        | 1,962.890 μs | 7.8197 μs | 6.9320 μs | 1,963.155 μs |  1.00 |    0.00 | 0.0153 | 0.0153 |      64 B |        1.00 |



</details>

Reply:
> For performance purposes, I could potentially only perform the sort if the ZipArchive is in Update mode. There are some cases where users would see entries appear in a different order depending on whether they opened it in Read or Update mode, but that might not be a problem - they're usually going to open it by name.

I like this suggestion. If this new behavior would also affect Write mode, it's not really needed there. I would welcome making this behavior exclusive to Update mode. Can you please make that change?

> I don't think we've got a unit test where the order of central directory entries is different to the order of local file headers, I'll add that over the next few days.

Thanks, yes. Please add one. Let me know if I can assist here.

Reply:
Sounds good, thanks @carlossanlop. I've added the condition, and both cases are tested by the `ArchivesInOffsetOrder_UpdateMode` and `ArchivesInCentralDirectoryOrder_ReadMode` methods.